### PR TITLE
FOGL-1533/1496/1512 - Configuration key may contain spaces and other characters also + add_service() refactored

### PR DESCRIPTION
**Pull request by amarendra-dianomic**

### CODE REVIEW

**Comment by praveen-garg** [comment]
Needs a through review!

File: python/foglamp/services/core/api/service.py [hunk 1]
```diff
@@ -68,21 +68,18 @@ def get_service_records():
     Create a new service to run a specific plugin
 
     :Example:
-             curl -X POST /foglamp/service -d '{"name": "furnace4", "type": "south", "plugin": "dht11"}'
+             curl -X POST http://localhost:8081/foglamp/service -d '{"type": "south", "plugin": "dht11", "enabled": true}'
```
Comment (MarkRiddoch) [inline python/foglamp/services/core/api/service.py]: No, this is completely wrong. The name is the important thing and the thing that defines a service, not the plugin. The same plugin maybe used by multiple services. Please remove this change.

Comment (praveen-garg) [reply python/foglamp/services/core/api/service.py]: yes! `name` is required. See how `FogLAMP Storage` and `FogLAMP Core` got the names. the problem FOGL-1533 describes is, don;t consider the service name as plugin config to load!

**Comment by amarendra-dianomic** [comment]
@MarkRiddoch @praveen-garg yes, removal of **name** param was incorrect. This has been fixed now. Also, making provision of space and other characters in **name** required changes at so many other places, notably in scheduler.py and process.py and configuration manager API. This of course requires rather thorough manual testing and updating unit tests to include checking for new values in **name** field. Updating unit tests may be taken up via another ticket.

**Comment by amarendra-dianomic** [comment]
@MarkRiddoch I have partially implemented #908 for FOGL-1520 in this PR and this removed the requirement of changes in scheduler.py and process.py. I have checked for South services and everything seems to be working fine. 

Service creation, with spaces in task/service name, is not working in #908 by itself. So #908 for FOGL-1520 alone is not sufficient for allowing spaces in task/service names. In fact, both #915 (this PR) and #908 together are required for allowing spaces in task/service names.

File: python/foglamp/common/microservice_management_client/microservice_management_client.py [hunk 2]
```diff
@@ -34,7 +35,10 @@ def register_service(self, service_registration_payload):
         management interface for that microservice
         :return: a JSON object containing the UUID of the newly registered service
         """
-        self._management_client_conn.request(method='POST', url='/foglamp/service', body=json.dumps(service_registration_payload))
+        url = '/foglamp/service'
+        url = urllib.parse.quote(url)
```
Comment (praveen-garg) [inline python/foglamp/common/microservice_management_client/microservice_management_client.py]: Why this is needed?

File: python/foglamp/common/microservice_management_client/microservice_management_client.py [hunk 3]
```diff
@@ -63,7 +67,10 @@ def unregister_service(self, microservice_id):
         :param microservice_id: string UUID of microservice
         :return: a JSON object containing the UUID of the unregistered service
         """
-        self._management_client_conn.request(method='DELETE', url='/foglamp/service/{}'.format(microservice_id))
+        url = '/foglamp/service/{}'.format(microservice_id)
+        url = urllib.parse.quote(url)
```
Comment (praveen-garg) [inline python/foglamp/common/microservice_management_client/microservice_management_client.py]: why this is needed, `microservice_id` is a uuid string here

File: python/foglamp/common/microservice_management_client/microservice_management_client.py [hunk 4]
```diff
@@ -91,8 +98,11 @@ def register_interest(self, category, microservice_id):
         :return: A JSON object containing a registration ID for this registration
         """
 
+        url = '/foglamp/interest'
+        url = urllib.parse.quote(url)
```
Comment (praveen-garg) [inline python/foglamp/common/microservice_management_client/microservice_management_client.py]: why this is needed?

File: python/foglamp/common/microservice_management_client/microservice_management_client.py [hunk 5]
```diff
@@ -118,7 +128,10 @@ def unregister_interest(self, registered_interest_id):
         :param registered_interest_id: registered interest id for a configuration category
         :return: A JSON object containing the unregistered interest id
         """
-        self._management_client_conn.request(method='DELETE', url='/foglamp/interest/{}'.format(registered_interest_id))
+        url = '/foglamp/interest/{}'.format(registered_interest_id)
+        url = urllib.parse.quote(url)
```
Comment (praveen-garg) [inline python/foglamp/common/microservice_management_client/microservice_management_client.py]: why this is needed, `registered_interest_id` is uuid str here

File: python/foglamp/common/microservice_management_client/microservice_management_client.py [hunk 6]
```diff
@@ -224,6 +242,7 @@ def create_configuration_category(self, category_data):
         :return:
         """
         url = '/foglamp/service/category'
+        url = urllib.parse.quote(url)
```
Comment (praveen-garg) [inline python/foglamp/common/microservice_management_client/microservice_management_client.py]: why this is needed here?

File: python/foglamp/services/core/api/service.py [hunk 7]
```diff
@@ -95,46 +96,50 @@ def get_service_records():
         is_enabled = True if ((type(enabled) is str and enabled.lower() in ['t', 'true']) or (
             (type(enabled) is bool and enabled is True))) else False
 
-        storage = connect.get_storage_async()
+        # ---------------------------------------- Sanity checks
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: is this the change for other ticket?

File: python/foglamp/services/core/api/service.py [hunk 8]
```diff
@@ -95,46 +96,50 @@ def get_service_records():
...
+            plugin_module_path = "foglamp.plugins.south" if service_type == 'south' else \
+                                 "foglamp.plugins.north" if service_type == 'north' else None
+            import_file_name = "{path}.{dir}.{file}".format(path=plugin_module_path, dir=plugin, file=plugin)
+            _plugin = __import__(import_file_name, fromlist=[''])
+
+            # Fetch configuration from the configuration defined in the plugin
+            plugin_info = _plugin.plugin_info()
+            plugin_config = plugin_info['config']
+        except ImportError as ex:
+            raise web.HTTPInternalServerError(reason='Plugin "{}" import problem from path "{}". {}'.format(plugin, plugin_module_path, str(ex)))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: it should be HTTP not found, even if this change is in via this PR.

File: python/foglamp/services/core/api/service.py [hunk 9]
```diff
@@ -95,46 +96,50 @@ def get_service_records():
...
             raise web.HTTPBadRequest(reason='A service with that name already exists')
 
         # Check that the schedule name is not already registered
-        payload = PayloadBuilder().SELECT("schedule_name").WHERE(['schedule_name', '=', name]).payload()
-        result = await storage.query_tbl_with_payload('schedules', payload)
-        count = result['count']
+        count = await check_schedules(name)
         if count != 0:
             raise web.HTTPBadRequest(reason='A schedule with that name already exists')
+        # ----------------------------------------
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: remove this  line

File: python/foglamp/services/core/api/service.py [hunk 10]
```diff
@@ -143,24 +148,59 @@ def get_service_records():
                                              category_value=plugin_config,
                                              keep_original_items=True)
         except ImportError as ex:
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: is this still an ImportError?

File: python/foglamp/services/core/api/service.py [hunk 11]
```diff
@@ -143,24 +148,59 @@ def get_service_records():
...
+    return result['count']
+
+async def check_scheduled_processes(process_name):
+    storage = connect.get_storage_async()
+    payload = PayloadBuilder().SELECT("schedule_name").WHERE(['schedule_name', '=', process_name]).payload()
+    result = await storage.query_tbl_with_payload('schedules', payload)
+    return result['count']
+
+async def revert_scheduled_processes(process_name):
+    return
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: return and do nothing?

File: python/foglamp/services/core/api/service.py [hunk 12]
```diff
@@ -143,24 +148,59 @@ def get_service_records():
...
+    storage = connect.get_storage_async()
+    payload = PayloadBuilder().WHERE(['name', '=', process_name]).payload()
+    result = await storage.delete_from_tbl('scheduled_processes', payload)
+
+async def revert_configuration(key):
+    return
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: return and do nothing?

File: python/foglamp/services/core/api/configuration.py [hunk 13]
```diff
@@ -131,8 +132,8 @@
     :Example:
             curl -X GET http://localhost:8081/foglamp/category/PURGE_READ/age
     """
-    category_name = request.match_info.get('category_name', None)
-    config_item = request.match_info.get('config_item', None)
+    category_name = urllib.parse.unquote(request.match_info.get('category_name', None))
+    config_item = urllib.parse.unquote(request.match_info.get('config_item', None))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/configuration.py]: I'm not sure about these 2: `category_name` and `config_item`

**Comment by amarendra-dianomic** [comment]
@MarkRiddoch yes, this PR contains solution for FOGL-1496/1512 also and I have rebased it with develop.

File: python/foglamp/services/core/api/service.py [hunk 14]
```diff
@@ -95,72 +96,105 @@ def get_service_records():
...
+async def check_scheduled_processes(process_name):
+    storage = connect.get_storage_async()
+    payload = PayloadBuilder().SELECT("schedule_name").WHERE(['schedule_name', '=', process_name]).payload()
+    result = await storage.query_tbl_with_payload('schedules', payload)
+    return result['count']
+
+async def revert_scheduled_processes(process_name):
+    storage = connect.get_storage_async()
+    payload = PayloadBuilder().WHERE(['name', '=', process_name]).payload()
+    result = await storage.delete_from_tbl('scheduled_processes', payload)
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: its result does not matter? as it has no return!

File: python/foglamp/services/core/api/service.py [hunk 15]
```diff
@@ -95,72 +96,105 @@ def get_service_records():
...
+
+async def revert_configuration(key):
+    storage = connect.get_storage_async()
+    payload = PayloadBuilder().WHERE(['key', '=', key]).payload()
+    result = await storage.delete_from_tbl('configuration', payload)
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: its result does not matter? as it has no return. Also please a line-break at the end of the file

File: python/foglamp/services/core/api/service.py (hunk 7 above)
Comment (praveen-garg) [reply python/foglamp/services/core/api/service.py]: Please remove this `# ---------------------------------------- Sanity checks`

File: python/foglamp/common/microservice_management_client/microservice_management_client.py [hunk 16]
```diff
@@ -147,10 +156,10 @@ def get_services(self, service_name=None, service_type=None):
         url = '/foglamp/service'
         delimeter = '?'
         if service_name:
-            url = '{}{}name={}'.format(url, delimeter, service_name)
+            url = '{}{}name={}'.format(url, delimeter, urllib.parse.quote(service_name))
             delimeter = '&'
         if service_type:
-            url = '{}{}type={}'.format(url, delimeter, service_type)
+            url = '{}{}type={}'.format(url, delimeter, urllib.parse.quote(service_type))
```
Comment (praveen-garg) [inline python/foglamp/common/microservice_management_client/microservice_management_client.py]: service_type can NOT have space

File: python/foglamp/services/core/api/configuration.py [hunk 17]
```diff
@@ -59,7 +60,7 @@
     :Example:
             curl -X GET http://localhost:8081/foglamp/category/PURGE_READ
     """
-    category_name = request.match_info.get('category_name', None)
+    category_name = urllib.parse.unquote(request.match_info.get('category_name', None))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/configuration.py]: what will be the value of `urllib.parse.unquote(None)` ?

File: python/foglamp/services/core/api/configuration.py [hunk 18]
```diff
@@ -159,8 +160,8 @@
         curl -X PUT -H "Content-Type: application/json" -d '{"value": 24}' http://localhost:8081/foglamp/category/PURGE_READ/age
 
     """
-    category_name = request.match_info.get('category_name', None)
-    config_item = request.match_info.get('config_item', None)
+    category_name = urllib.parse.unquote(request.match_info.get('category_name', None))
+    config_item = urllib.parse.unquote(request.match_info.get('config_item', None))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/configuration.py]: as above!

File: python/foglamp/services/core/api/configuration.py [hunk 19]
```diff
@@ -196,8 +197,8 @@
         curl -d '{"default": "true", "description": "Test description", "type": "boolean"}' -X POST https://localhost:1995/foglamp/category/{category_name}/{new_config_item} --insecure
         curl -d '{"default": "true", "description": "Test description", "type": "boolean", "value": "false"}' -X POST https://localhost:1995/foglamp/category/{category_name}/{new_config_item} --insecure
     """
-    category_name = request.match_info.get('category_name', None)
-    new_config_item = request.match_info.get('config_item', None)
+    category_name = urllib.parse.unquote(request.match_info.get('category_name', None))
+    new_config_item = urllib.parse.unquote(request.match_info.get('config_item', None))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/configuration.py]: as above!

File: python/foglamp/services/core/api/configuration.py [hunk 20]
```diff
@@ -266,8 +267,8 @@
         curl -X DELETE http://localhost:8081/foglamp/category/PURGE_READ/age/value
 
     """
-    category_name = request.match_info.get('category_name', None)
-    config_item = request.match_info.get('config_item', None)
+    category_name = urllib.parse.unquote(request.match_info.get('category_name', None))
+    config_item = urllib.parse.unquote(request.match_info.get('config_item', None))
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/configuration.py]: as above!

File: python/foglamp/services/core/api/service.py [hunk 21]
```diff
@@ -95,72 +96,105 @@ def get_service_records():
...
             (type(enabled) is bool and enabled is True))) else False
 
-        storage = connect.get_storage_async()
+        # ---------------------------------------- Sanity checks
+        # Check if a valid plugin has been provided
+        try:
+            # "plugin_module_path" is fixed by design. It is MANDATORY to keep the plugin in the exactly similar named
+            # folder, within the plugin_module_path.
+            plugin_module_path = "foglamp.plugins.south" if service_type == 'south' else \
+                                 "foglamp.plugins.north" if service_type == 'north' else None
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: it should NEVER be none! better to check `service_type in ['south', 'north']` and then `plugin_module_path = "foglamp.plugins.{}".format(service_type)`

File: python/foglamp/services/core/api/service.py [hunk 22]
```diff
@@ -95,72 +96,105 @@ def get_service_records():
...
-        if service_type == 'south':
-            script = '["services/south"]'
-            plugin_module_path = "foglamp.plugins.south"
-        if service_type == 'north':
-            script = '["services/north"]'
-            plugin_module_path = "foglamp.plugins.north"
+        # Now first create the scheduled process entry for the new service
+        storage = connect.get_storage_async()
+        script = '["services/south"]' if service_type == 'south' else \
+                 '["services/north"]' if service_type == 'north' else None
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: as above! it should NEVER be None!

Comment (praveen-garg) [reply python/foglamp/services/core/api/service.py]: why we should make a call to payloadbuilder + storage with None!

File: python/foglamp/services/core/api/service.py [hunk 23]
```diff
@@ -95,72 +96,105 @@ def get_service_records():
...
+            raise web.HTTPInternalServerError(reason='Failed to created schedule. {}'.format(str(ins_ex)))
 
         return web.json_response({'name': name, 'id': str(schedule.schedule_id)})
 
     except ValueError as ex:
         raise web.HTTPNotFound(reason=str(ex))
 
+
+async def check_schedules(schedule_name):
+    storage = connect.get_storage_async()
```
Comment (praveen-garg) [inline python/foglamp/services/core/api/service.py]: please do this once is caller def, and pass to this and `check_scheduled_processes` and `revert_scheduled_processes` as needed!

**Comment by praveen-garg** [comment]
Please fix the unit tests!

File: python/foglamp/common/microservice_management_client/microservice_management_client.py (hunk 16 above)
Comment (praveen-garg) [reply python/foglamp/common/microservice_management_client/microservice_management_client.py]: don't allow, Please see:

```
class ServiceRecord(object):
    """Used to information regarding a registered microservice.
    """

    class Type(IntEnum):
        """Enumeration for Service Types"""

        Storage = 1
        Core = 2
        Southbound = 3
```

**Comment by amarendra-dianomic** [comment]
@praveen-garg feedback changes done. Pl check.

**Comment by praveen-garg** [comment]
restrict service name, category name to have `/` character as it may manipulate url.

Also see `:` works?

**Pull request merged**